samples/nearby-chat/src/main/java/org/mozilla/samples/nearby/chat/MainActivity.kt

```diff
@@ -0,0 +1,159 @@
+    // initialized in init() after permissions granted
+    private lateinit var connection: NearbyConnection
+
+    // updated in listener
+    private lateinit var state: ConnectionState
+
+    // Lifecycle methods
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_main)
+
+        // This button must not be enabled until permissions have been granted.
+        resetButton.setOnClickListener {
+            connection.disconnect()
+            init()
+        }
+
+        // These two buttons must not be enabled until connection is initialized in init().
+        advertiseButton.setOnClickListener {
+            disableConnectionButtons()
+            connection.startAdvertising()
+        }
+        discoverButton.setOnClickListener {
+            disableConnectionButtons()
+            connection.startDiscovering()
+        }
+
+        // This button must not be enabled unless the current state is READY_TO_SEND.
+        sendMessageButton.setOnClickListener {
+            if (state !is ConnectionState.ReadyToSend) {
+                Logger.error("Attempt to send message when in state ${state.name}")
+            } else {
+                connection.sendMessage(outgoingMessageText.text.toString())
+                outgoingMessageText.text?.clear()
+            }
+        }
+    }
+
+    override fun onStart() {
+        super.onStart()
+        requestPermissions()
+    }
+
+    override fun onStop() {
+        super.onStop()
+        // Unless permissions were granted and init() ran to completion, connection may be uninitialized.
+        if (this::connection.isInitialized) {
+            connection.disconnect()
+        }
+    }
+
+    // Permissions
+
+    private fun requestPermissions() {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            val neededPermissions = NearbyConnection.PERMISSIONS.filter {
+                ContextCompat.checkSelfPermission(this, it) != PackageManager.PERMISSION_GRANTED
+            }
+
+            if (neededPermissions.isNotEmpty()) {
+                ActivityCompat.requestPermissions(this, neededPermissions.toTypedArray(), 0)
+            } else {
+                init()
+            }
+        }
+    }
+
+    override fun onRequestPermissionsResult(
+        requestCode: Int,
+        permissions: Array<String>,
+        grantResults: IntArray
+    ) {
+        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
+
+        if (grantResults.any { it == PackageManager.PERMISSION_DENIED }) {
+            statusText.text = getString(R.string.mozac_samples_nearby_chat_permissions_denied)
+            init()
+        }
+    }
+
+    // Called after permissions first granted or reset button pressed
+    private fun init() {
+        // Can't do in onCreate because context not ready.
+        connection = NearbyConnection(this)
+        connection.listener =
```

Comment:
I highly recommend looking into the ObserverRegistry for the NearbyConnection so that you can observe different state changes so you don't have to handle all UI/user interactions in one method.

Reply:
I've switched to using `ObserverRegistry`, but isn't that orthogonal to implementing `NearbyConnectionListener`/`NearbyConnectionObserver` in an anonymous class declared within a single `init()` method, or am I misunderstanding?

Reply:
In any case, I've made it a named inner class.

Reply:
and the listener/observer is no longer passed in to the constructor.